File: rustspider/src/checkpoint_test.rs
//! RustSpider 断点续爬模块测试

use crate::checkpoint::{CheckpointManager, CheckpointState};
use std::collections::HashMap;
use std::time::Duration;

#[test]
fn test_checkpoint_state_new() {
    let visited = vec!["url1".to_string(), "url2".to_string()];
    let pending = vec!["url3".to_string()];
    let mut stats = HashMap::new();
    stats.insert("total".to_string(), serde_json::json!(100));
    
    let state = CheckpointState::new(
        "test_spider".to_string(),
        visited,
        pending,
        stats,
        HashMap::new(),
    );
    
    assert_eq!(state.spider_id, "test_spider");
    assert_eq!(state.visited_urls.len(), 2);
    assert_eq!(state.pending_urls.len(), 1);
    assert!(!state.checksum.is_empty());
}

#[test]
fn test_checkpoint_state_checksum() {
    let mut stats1 = HashMap::new();
    stats1.insert("total".to_string(), serde_json::json!(100));
    
    let state1 = CheckpointState::new(
        "test".to_string(),
        vec!["url1".to_string()],
        vec![],
        stats1,
        HashMap::new(),
    );
    
    let mut stats2 = HashMap::new();
    stats2.insert("total".to_string(), serde_json::json!(100));
    
    let state2 = CheckpointState::new(
        "test".to_string(),
        vec!["url1".to_string()],
        vec![],
        stats2,
        HashMap::new(),
    );
    
    // 相同状态应该有相同校验和
    assert_eq!(state1.checksum, state2.checksum);
    
    // 不同状态应该有不同校验和
    let mut stats3 = HashMap::new();
    stats3.insert("total".to_string(), serde_json::json!(200));
    
    let state3 = CheckpointState::new(
        "test".to_string(),
        vec!["url1".to_string()],
        vec![],
        stats3,
        HashMap::new(),
    );
    
    assert_ne!(state1.checksum, state3.checksum);
}

#[test]
fn test_checkpoint_state_verify() {
    let mut stats = HashMap::new();
    stats.insert("total".to_string(), serde_json::json!(100));
    
    let state = CheckpointState::new(
        "test".to_string(),
        vec!["url1".to_string()],
        vec![],
        stats,
        HashMap::new(),
    );
    
    assert!(state.verify_checksum());
    
    // 篡改校验和
    let mut bad_state = state.clone();
    bad_state.checksum = "bad_checksum".to_string();
    
    assert!(!bad_state.verify_checksum());
}

#[test]
fn test_checkpoint_manager_new() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), Some(300));
    
    assert!(temp_dir.path().exists());
}

#[test]
fn test_checkpoint_manager_save_and_load() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let mut stats = HashMap::new();
    stats.insert("total".to_string(), serde_json::json!(100));
    
    // 保存
    checkpoint.save(
        "test_spider",
        vec!["url1".to_string(), "url2".to_string()],
        vec!["url3".to_string()],
        stats.clone(),
        HashMap::new(),
        true,
    ).unwrap();
    
    // 加载
    let state = checkpoint.load("test_spider").unwrap();
    
    assert_eq!(state.spider_id, "test_spider");
    assert_eq!(state.visited_urls.len(), 2);
    assert_eq!(state.pending_urls.len(), 1);
}

#[test]
fn test_checkpoint_manager_save_cached() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    // 保存到缓存（不立即保存）
    checkpoint.save(
        "test_cached",
        vec![],
        vec![],
        HashMap::new(),
        HashMap::new(),
        false,
    ).unwrap();
    
    // 从缓存加载
    let state = checkpoint.load("test_cached").unwrap();
    assert!(state.is_some());
}

#[test]
fn test_checkpoint_manager_load_nonexistent() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let state = checkpoint.load("nonexistent");
    assert!(state.is_none());
}

#[test]
fn test_checkpoint_manager_delete() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    // 保存
    checkpoint.save(
        "test_delete",
        vec![],
        vec![],
        HashMap::new(),
        HashMap::new(),
        true,
    ).unwrap();
    
    // 删除
    checkpoint.delete("test_delete").unwrap();
    
    // 加载应该返回 None
    let state = checkpoint.load("test_delete");
    assert!(state.is_none());
}

#[test]
fn test_checkpoint_manager_list() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    // 保存多个
    for i in 0..3 {
        checkpoint.save(
            &format!("spider_{}", i),
            vec![],
            vec![],
            HashMap::new(),
            HashMap::new(),
            true,
        ).unwrap();
    }
    
    let checkpoints = checkpoint.list_checkpoints();
    
    assert_eq!(checkpoints.len(), 3);
    assert!(checkpoints.contains(&"spider_0".to_string()));
}

#[test]
fn test_checkpoint_manager_get_stats() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let mut stats = HashMap::new();
    stats.insert("total".to_string(), serde_json::json!(100));
    
    checkpoint.save(
        "test_stats",
        vec!["url1".to_string(), "url2".to_string()],
        vec!["url3".to_string()],
        stats,
        HashMap::new(),
        true,
    ).unwrap();
    
    let result_stats = checkpoint.get_stats("test_stats").unwrap();
    
    assert_eq!(result_stats["visited_count"], serde_json::json!(2));
    assert_eq!(result_stats["pending_count"], serde_json::json!(1));
}

#[test]
fn test_checkpoint_manager_get_stats_nonexistent() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let stats = checkpoint.get_stats("nonexistent");
    assert!(stats.is_none());
}

#[test]
fn test_checkpoint_manager_auto_save() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), Some(1));
    
    // 保存到缓存（不立即保存）
    checkpoint.save(
        "test_auto",
        vec!["url1".to_string()],
        vec![],
        HashMap::new(),
        HashMap::new(),
        false,
    ).unwrap();
    
    // 等待自动保存
    std::thread::sleep(Duration::from_millis(1500));
    
    // 检查文件是否存在
    let file_path = temp_dir.path().join("test_auto.checkpoint.json");
    assert!(file_path.exists(), "自动保存应该创建文件");
}

#[test]
fn test_checkpoint_manager_close() {
    let temp_dir = tempfile::tempdir().unwrap();
    
    {
        let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
        
        // 保存到缓存（不立即保存）
        checkpoint.save(
            "test_close",
            vec!["url1".to_string()],
            vec![],
            HashMap::new(),
            HashMap::new(),
            false,
        ).unwrap();
        
        // checkpoint 在这里被 drop，应该保存所有缓存
    }
    
    // 检查文件是否存在
    let file_path = temp_dir.path().join("test_close.checkpoint.json");
    assert!(file_path.exists(), "关闭时应该保存所有缓存状态");
}

#[test]
fn test_checkpoint_manager_concurrent_save() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = std::sync::Arc::new(CheckpointManager::new(temp_dir.path().to_str().unwrap(), None));
    
    let mut handles = vec![];
    
    // 并发保存
    for i in 0..5 {
        let checkpoint = std::sync::Arc::clone(&checkpoint);
        let handle = std::thread::spawn(move || {
            checkpoint.save(
                &format!("spider_{}", i),
                vec![format!("url_{}", i)],
                vec![],
                HashMap::new(),
                HashMap::new(),
                true,
            ).unwrap();
        });
        handles.push(handle);
    }
    
    // 等待所有线程完成
    for handle in handles {
        handle.join().unwrap();
    }
    
    // 验证所有 checkpoint 都被保存
    let checkpoints = checkpoint.list_checkpoints();
    assert_eq!(checkpoints.len(), 5);
}

#[test]
fn test_checkpoint_manager_large_data() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    // 创建 1000 个 URL
    let visited_urls: Vec<String> = (0..1000)
        .map(|i| format!("http://example.com/page{}", i))
        .collect();
    
    checkpoint.save(
        "test_large",
        visited_urls,
        vec![],
        HashMap::new(),
        HashMap::new(),
        true,
    ).unwrap();
    
    let state = checkpoint.load("test_large").unwrap();
    assert_eq!(state.visited_urls.len(), 1000);
}

#[test]
fn test_checkpoint_manager_special_characters() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let visited_urls = vec![
        "http://example.com/page?param=value&other=123".to_string(),
        "http://example.com/page with spaces".to_string(),
        "http://example.com/page/中文/unicode".to_string(),
    ];
    
    checkpoint.save(
        "test_special",
        visited_urls,
        vec![],
        HashMap::new(),
        HashMap::new(),
        true,
    ).unwrap();
    
    let state = checkpoint.load("test_special").unwrap();
    assert_eq!(state.visited_urls.len(), 3);
}

#[test]
fn test_checkpoint_manager_empty_state() {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    checkpoint.save(
        "test_empty",
        vec![],
        vec![],
        HashMap::new(),
        HashMap::new(),
        true,
    ).unwrap();
    
    let state = checkpoint.load("test_empty").unwrap();
    assert_eq!(state.visited_urls.len(), 0);
    assert_eq!(state.pending_urls.len(), 0);
}

#[tokio::test]
async fn test_checkpoint_manager_drop() {
    let temp_dir = tempfile::tempdir().unwrap();
    
    {
        let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
        
        checkpoint.save(
            "test_drop",
            vec!["url1".to_string()],
            vec![],
            HashMap::new(),
            HashMap::new(),
            false,
        ).unwrap();
        
        // checkpoint 在这里被 drop
    }
    
    // 检查文件是否存在
    let file_path = temp_dir.path().join("test_drop.checkpoint.json");
    assert!(file_path.exists());
}

#[bench]
fn bench_checkpoint_save(b: &mut test::Bencher) {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    let mut stats = HashMap::new();
    stats.insert("total".to_string(), serde_json::json!(100));
    
    b.iter(|| {
        checkpoint.save(
            "benchmark",
            vec!["url1".to_string()],
            vec![],
            stats.clone(),
            HashMap::new(),
            true,
        ).unwrap();
    });
}

#[bench]
fn bench_checkpoint_load(b: &mut test::Bencher) {
    let temp_dir = tempfile::tempdir().unwrap();
    let checkpoint = CheckpointManager::new(temp_dir.path().to_str().unwrap(), None);
    
    checkpoint.save(
        "benchmark",
        vec!["url1".to_string()],
        vec![],
        HashMap::new(),
        HashMap::new(),
        true,
    ).unwrap();
    
    b.iter(|| {
        checkpoint.load("benchmark");
    });
}

File: gospider/core/checkpoint_test.go
package core

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

// TestCheckpointManagerNew - 测试创建管理器
func TestCheckpointManagerNew(t *testing.T) {
	tempDir := t.TempDir()

	manager, err := NewCheckpointManager(tempDir, time.Second*5, 10)
	if err != nil {
		t.Fatal(err)
	}
	if manager == nil {
		t.Fatal("CheckpointManager 不应该为 nil")
	}

	// 验证目录创建
	if _, err := os.Stat(tempDir); os.IsNotExist(err) {
		t.Error("checkpoint 目录应该被创建")
	}

	manager.Close()
}

// TestCheckpointSaveAndLoad - 测试保存和加载
func TestCheckpointSaveAndLoad(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_spider"
	visitedURLs := []string{"http://a.com", "http://b.com"}
	pendingURLs := []string{"http://c.com"}
	stats := map[string]interface{}{
		"total":   100,
		"success": 95,
	}
	config := map[string]interface{}{
		"threads": 10,
	}

	// 保存
	err = manager.Save(spiderID, visitedURLs, pendingURLs, stats, config, true)
	if err != nil {
		t.Fatalf("保存失败：%v", err)
	}

	// 加载
	state, err := manager.Load(spiderID)
	if err != nil {
		t.Fatalf("加载失败：%v", err)
	}

	if state == nil {
		t.Fatal("state 不应该为 nil")
	}

	// 验证数据
	if state.SpiderID != spiderID {
		t.Errorf("SpiderID 不匹配：期望 %s, 得到 %s", spiderID, state.SpiderID)
	}

	if len(state.VisitedURLs) != 2 {
		t.Errorf("VisitedURLs 数量错误：期望 2, 得到 %d", len(state.VisitedURLs))
	}

	if len(state.PendingURLs) != 1 {
		t.Errorf("PendingURLs 数量错误：期望 1, 得到 %d", len(state.PendingURLs))
	}
}

// TestCheckpointSaveCached - 测试缓存保存
func TestCheckpointSaveCached(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_cached"

	// 保存到缓存（不立即保存）
	err = manager.Save(spiderID, []string{}, []string{}, map[string]interface{}{}, map[string]interface{}{}, false)
	if err != nil {
		t.Fatalf("保存失败：%v", err)
	}

	// 文件不应该存在
	filePath := filepath.Join(tempDir, spiderID+".checkpoint.json")
	if _, err := os.Stat(filePath); err == nil {
		t.Error("文件不应该存在（因为没立即保存）")
	}

	// 但从缓存应该能加载
	state, err := manager.Load(spiderID)
	if err != nil {
		t.Fatalf("加载失败：%v", err)
	}

	if state == nil {
		t.Error("state 不应该为 nil")
	}
}

// TestCheckpointLoadNonexistent - 测试加载不存在
func TestCheckpointLoadNonexistent(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	state, err := manager.Load("nonexistent")
	if err != nil {
		t.Fatalf("加载不应该失败：%v", err)
	}

	if state != nil {
		t.Error("state 应该为 nil")
	}
}

// TestCheckpointDelete - 测试删除
func TestCheckpointDelete(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_delete"

	// 先保存
	manager.Save(spiderID, []string{}, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)

	// 删除
	err = manager.Delete(spiderID)
	if err != nil {
		t.Fatalf("删除失败：%v", err)
	}

	// 验证文件不存在
	filePath := filepath.Join(tempDir, spiderID+".checkpoint.json")
	if _, err := os.Stat(filePath); err == nil {
		t.Error("文件应该被删除")
	}
}

// TestCheckpointList - 测试列出 checkpoint
func TestCheckpointList(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	// 保存多个
	for i := 0; i < 3; i++ {
		spiderID := "spider_" + string(rune('0'+i))
		manager.Save(spiderID, []string{}, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)
	}

	checkpoints, err := manager.ListCheckpoints()
	if err != nil {
		t.Fatalf("列出失败：%v", err)
	}

	if len(checkpoints) != 3 {
		t.Errorf("checkpoint 数量错误：期望 3, 得到 %d", len(checkpoints))
	}
}

// TestCheckpointGetStats - 测试获取统计
func TestCheckpointGetStats(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_stats"
	stats := map[string]interface{}{
		"total":   100,
		"success": 95,
	}

	manager.Save(spiderID, []string{"url1", "url2"}, []string{"url3"}, stats, map[string]interface{}{}, true)

	resultStats, err := manager.GetStats(spiderID)
	if err != nil {
		t.Fatalf("获取统计失败：%v", err)
	}

	if resultStats == nil {
		t.Fatal("resultStats 不应该为 nil")
	}

	visitedCount, ok := resultStats["visited_count"].(int)
	if !ok || visitedCount != 2 {
		t.Errorf("visited_count 错误：期望 2, 得到 %v", visitedCount)
	}
}

// TestCheckpointAutoSave - 测试自动保存
func TestCheckpointAutoSave(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, time.Second, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_auto"

	// 保存到缓存（不立即保存）
	manager.Save(spiderID, []string{"url1"}, []string{}, map[string]interface{}{}, map[string]interface{}{}, false)

	// 等待自动保存
	time.Sleep(time.Second * 2)

	// 检查文件是否被创建
	filePath := filepath.Join(tempDir, spiderID+".checkpoint.json")
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		t.Error("自动保存应该创建文件")
	}
}

// TestCheckpointChecksum - 测试校验和
func TestCheckpointChecksum(t *testing.T) {
	state := &CheckpointState{
		SpiderID:    "test",
		Timestamp:   "2026-03-23T10:00:00Z",
		VisitedURLs: []string{"url1"},
		PendingURLs: []string{},
		Stats:       map[string]interface{}{},
		Config:      map[string]interface{}{},
	}

	checksum1 := state.ComputeChecksum()
	checksum2 := state.ComputeChecksum()

	if checksum1 != checksum2 {
		t.Error("相同状态应该有相同校验和")
	}

	// 修改状态
	state.VisitedURLs = append(state.VisitedURLs, "url2")
	checksum3 := state.ComputeChecksum()

	if checksum1 == checksum3 {
		t.Error("不同状态应该有不同校验和")
	}
}

// TestCheckpointSaveFromSpider - 测试从当前主线爬虫保存
func TestCheckpointSaveFromSpider(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	config := DefaultSpiderConfig()
	config.Name = "test_spider"
	spider := NewSpider(config)

	err = manager.SaveFromSpider(spider, true)
	if err != nil {
		t.Fatalf("保存失败：%v", err)
	}

	state, err := manager.Load("test_spider")
	if err != nil {
		t.Fatalf("加载失败：%v", err)
	}

	if state == nil {
		t.Error("state 不应该为 nil")
	}
}

// TestCheckpointConcurrentSave - 测试并发保存
func TestCheckpointConcurrentSave(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	done := make(chan bool, 5)

	// 并发保存
	for i := 0; i < 5; i++ {
		go func(index int) {
			spiderID := "spider_" + string(rune('0'+index))
			manager.Save(spiderID, []string{"url"}, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)
			done <- true
		}(i)
	}

	// 等待所有 goroutine 完成
	for i := 0; i < 5; i++ {
		<-done
	}

	// 验证
	checkpoints, _ := manager.ListCheckpoints()
	if len(checkpoints) != 5 {
		t.Errorf("checkpoint 数量错误：期望 5, 得到 %d", len(checkpoints))
	}
}

// TestCheckpointLargeData - 测试大数据保存
func TestCheckpointLargeData(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_large"
	visitedURLs := make([]string, 1000)
	for i := 0; i < 1000; i++ {
		visitedURLs[i] = "http://example.com/page" + string(rune(i))
	}

	err = manager.Save(spiderID, visitedURLs, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)
	if err != nil {
		t.Fatalf("保存失败：%v", err)
	}

	state, err := manager.Load(spiderID)
	if err != nil {
		t.Fatalf("加载失败：%v", err)
	}

	if len(state.VisitedURLs) != 1000 {
		t.Errorf("VisitedURLs 数量错误：期望 1000, 得到 %d", len(state.VisitedURLs))
	}
}

// TestCheckpointSpecialCharacters - 测试特殊字符
func TestCheckpointSpecialCharacters(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	spiderID := "test_special"
	visitedURLs := []string{
		"http://example.com/page?param=value&other=123",
		"http://example.com/page with spaces",
		"http://example.com/page/中文/unicode",
	}

	err = manager.Save(spiderID, visitedURLs, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)
	if err != nil {
		t.Fatalf("保存失败：%v", err)
	}

	state, err := manager.Load(spiderID)
	if err != nil {
		t.Fatalf("加载失败：%v", err)
	}

	if len(state.VisitedURLs) != 3 {
		t.Errorf("VisitedURLs 数量错误：期望 3, 得到 %d", len(state.VisitedURLs))
	}
}

// TestCheckpointClose - 测试关闭
func TestCheckpointClose(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		t.Fatal(err)
	}

	spiderID := "test_close"

	// 保存到缓存（不立即保存）
	manager.Save(spiderID, []string{"url1"}, []string{}, map[string]interface{}{}, map[string]interface{}{}, false)

	// 关闭（应该保存所有缓存）
	manager.Close()

	// 检查文件是否存在
	filePath := filepath.Join(tempDir, spiderID+".checkpoint.json")
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		t.Error("关闭时应该保存所有缓存状态")
	}
}

func TestCheckpointRetentionKeepsRecentHistory(t *testing.T) {
	tempDir := t.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 2)
	if err != nil {
		t.Fatal(err)
	}
	defer manager.Close()

	for i := 0; i < 3; i++ {
		err = manager.Save("history_spider", []string{"url"}, []string{}, map[string]interface{}{"seq": i}, map[string]interface{}{}, true)
		if err != nil {
			t.Fatalf("save failed: %v", err)
		}
		time.Sleep(10 * time.Millisecond)
	}

	history, err := filepath.Glob(filepath.Join(tempDir, "history_spider.checkpoint.*.json"))
	if err != nil {
		t.Fatalf("failed to list history files: %v", err)
	}
	if len(history) != 2 {
		t.Fatalf("expected 2 retained checkpoint versions, got %d (%v)", len(history), history)
	}
}

// BenchmarkCheckpointSave - 基准测试：保存
func BenchmarkCheckpointSave(b *testing.B) {
	tempDir := b.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		b.Fatal(err)
	}
	defer manager.Close()

	spiderID := "benchmark"
	visitedURLs := []string{"http://example.com/page"}
	stats := map[string]interface{}{"total": 100}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		manager.Save(spiderID, visitedURLs, []string{}, stats, map[string]interface{}{}, true)
	}
}

// BenchmarkCheckpointLoad - 基准测试：加载
func BenchmarkCheckpointLoad(b *testing.B) {
	tempDir := b.TempDir()
	manager, err := NewCheckpointManager(tempDir, 0, 10)
	if err != nil {
		b.Fatal(err)
	}
	defer manager.Close()

	spiderID := "benchmark"
	manager.Save(spiderID, []string{"url"}, []string{}, map[string]interface{}{}, map[string]interface{}{}, true)

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		manager.Load(spiderID)
	}
}
